README


        lpcisp -- NXP ISP programming tool

lpcisp is intended to be a clean and flexible way to identify, erase, read, and write
NXP LPC-family components which support serial ISP (In-System Programming) mode.

The basic usage is:
        lpcisp [options] device

The device argument is the serial device used to communicate with the
target processor.  All options are preceded by a dash.  To get the
current complete list of options invoke the program with no arguments.

Please note:
        1) on exit the target is reset.  If reset isn't mapped a SYSRESETREQ is invoked.
        	Due to some ambiguity as to what the boot loader manages to respond after
        	the "GO" command and before the part resets a couple of unexpected bytes
        	may be received by the host application on startup.
        2) the set of supported devices is incomplete and some may be inaccurate.  Adding
                more parts is simply a matter of adding lines to the list.  See partdesc.c.
                Please forward changes and additions via:
                   https://github.com/andrewpines/lpcisp/issues
        3) some devices are more finicky about the oscillator setting and baud rate than
                others.  if it has trouble communicating try reducing the baud rate.
        4) some devices (LPC2134, for instance) don't seem to communicate correctly unless
                the echo option is enabled.  The transfer is much faster with echo off, so
                it's off by default.  If the device identifies correctly but fails to
                acknowledge the transfer of blocks of data try enabling echo.

Supporting documentation:
        AN11229, UUencoding for UART ISP, Rev. 1 -- 22 June 2012
        UM10114, LPC21xx and LPC22xx User manual, Rev. 4 -- 2 May 2012
        UM10120, LPC2131/2/4/6/8 User manual, Rev. 4 -- 23 April 2012
        UM10211, LPC23xx User manual, Rev. 4.1 -- 5 September 2012
        UM10360, LPC176x/5x User manual, Rev. 4. 1 -- 19 December 2016
        UM10375, LPC1311/13/42/43 User manual, Rev. 5 -- 21 June 2012
        UM10398, LPC111x/LPC11Cxx User manual, Rev. 12.1, August 2013
        UM10430, LPC18xx User manual, Rev. 2.8 -- 10 December 2015
        UM10470, LPC178x/7x User manual, Rev. 3.1 -- 15 September 2014
        UM10601, LPC800 User Manaual, Rev. 1.3, 22 July 2013
        UM10736, LPC15xx User manual, Rev. 1.1 -- 3 March 2014
        UM10800, LPC82x User manual, Rev. 1.2 -- 5 October 2016
	UM10912, LPC546xx User manual, Rev. 1.9 -- 16 June 2017

================================================

COMMAND LINE OPTIONS

Following is a list of the available options:

-device name    specify an LPC device name (e.g., "LPC1311FHN33").  the name
                must exactly match one of the names listed by "lpcisp -parts".
                lpcisp will stop if the name specified does not match the name
                implied by the part ID code read from the device.  if no name
                is specified then no check is made and lpcisp will attempt to
                perform the specified operations on whatever device is detected.

-reset ctrl     map the reset function to a serial device signal.  ctrl may be
                one of dtr, ndtr, rts, or nrts.  lpcisp will assert (dtr, rts)
                or negate (ndtr, nrts) the selected signal to reset the target.
                default is none.

-isp ctrl       map the ISP function to a serial device signal.  ctrl may be
                one of dtr, ndtr, rts, or nrts.  lpcisp will assert (dtr, rts)
                or negate (ndtr, nrts) the selected signal to invoke ISP on the
                target.  Default is none.

-hold           if invoked the ISP signal will be held through the programming
                sequence.  this is for systems where the signal is used to take
                competing serial devices off the bus during programming.  Default
                is to negate ISP immediately after reset at start of programming
                sequence.

-baud rate      set the baud rate to be used for communicating with the target
                device.  Default is 115200.

-tbaud rate     set the baud rate to be used for terminal mode.  Default is 115200.

-clock rate     set the clock rate of the target device, in kHz.  Default is
                12000 (12MHz).

-retry num      set the number of times to attempt to sync to the target device.
                Default is 25.

-bin filename   load a binary file to be written to or compared against the flash
                memory of the target device.  Always loads at address 0x00000000.

-hex filename   load an intel hex file to be written to or compared against the flash
                memory of the target device.  Does not need to begin at address
                0x00000000 but if it is to be written to the target it does need
                to start on a page boundary.  This restriction may someday be lifted
                if an application appears where it would be useful.

-vector         LPC parts require vector 7 to be the 2's complement of the sum of
                vectors 0 through 6.  If the image starts at address 0x00000000 this
                is filled in automatically.  if the image does not start at 0x00000000
                but the beginning of the image is the vector table then this option
                will force the patching of vector 7 at the beginning of the image.

-erase          erase required sectors on target device (i.e., any sectors which
                coincide with the data about to be written).  Note that this may fail
                if the device is read-protected since it may not erase the sector
                with the code read protect word.

-eraseall       erase the target device (all sectors).

-echo           enable command echo (disabled by default).

-write          write the loaded file to flash on the target.

-verify         compare the loaded file with the flash on the target.  Note that this
                does not require writing the file first though if it is used without
                first writing the code it requires that code read protect is disabled.

-dump address length      dump a block of memory from the device.  Address may be anywhere
                within the device but it must be word-algned (multiple of four).  Length
                must also be word-aligned.  If this is executed without erase/write then
                code read protect must be disabled.

-term           after completion of other operations reset the target and drop to a
                terminal to communicate with it.  The baud rate is set by the -tbaud
                option.  Pressing <ESC> will reset the target (if -reset was configured)
                and pressing <CTRL-X> will exit.  All other key codes are passed to the
                target.

-verbose level  set the reporting level (0-4):
                  0: errors only.
                  1: minimum (only what was expressly requested, just memory dump).
                  2: information; basic information about devices and process.  This is the default.
                  3: debug process; details about the various tasks being executed.
                  4: debug full; complete dump of data exchanged with target.
                all messages at or below the selected level will be displayed (e.g., if level
                is set to 2 then messages at levels 0, 1, and 2 will be displayed).

-parts          dump a list of supported parts.

================================================

EXAMPLES

        ID
                The target processor can be identified without performing any operations on it.
                Assume:
                  the target has its reset signal connected to DTR
                  the target has its ISP signal connected to RTS
                  the serial device is /dev/ttyS0
                The command would be:
                  lpcisp -reset ndtr -isp nrts /dev/ttyS0

                Note that DTR and RTS are inverted.  This is the correct mapping if they are
                connected directly to the outputs of an RS232-to-logic level converter.

                if successful lpcisp will report something like the following:
                  device information:
                     part number:  LPC1112FHN33/[101|102]
                     ID:           0x2524d02b
                     flash size:   16384 bytes
                     SRAM size:    2048 bytes
                     UID:          f5000001 4ff53bb2 53581b51 0707212d
                     boot version: 1.7

                The ID, UID and boot version are read from the part.  The rest are looked up
                based on the ID.


        Erase
                To erase the device from the previous example the command would be:
                  lpcisp -reset ndtr -isp nrts -eraseall /dev/ttyS0


        Write
                The most common task is to write an image to a device.  To write an Intel
                hex file named image.hex to the target device in the above examples the
                command would be:
                  lpcisp -reset ndtr -isp nrts -hex image.hex -erase -write /dev/ttyS0


        Verify
                To verify the code after writing in the above example simply add the -verify
                option:
                  lpcisp -reset ndtr -isp nrts -hex image.hex -erase -write -verify /dev/ttyS0

                Alternatively, a previously-written image can be verified against a hex file
                if code protect is disabled.  This will report the result but will not alter
                the target device:
                  lpcisp -reset ndtr -isp nrts -hex image.hex -verify /dev/ttyS0


        Dump
                Similar to verify, a block of word-aligned memory can be dumped out.  To
                read 256 bytes starting at address 0x000200 from the target device in the
                previous examples the command would be:
                  lpcisp -reset ndtr -isp nrts -dump 0x200 256  /dev/ttyS0

                Note that the address and length arguments can be either decimal or hex
                (preceded by 0x).  Again, this only works after erasing the entire device
                or if code read protect is disabled.

        Terminal
                A serial terminal may be opened after other operations are complete.  Reset
                and ISP are left in their negated states.  The baud rate is set by the
                -tbaud option.  The terminal can be entered without other operations.  The
                command would be:
                  lpcisp -reset ndtr -isp nrts -tbaud 9600 -term /dev/ttyS0

================================================

LINKING AS A LIBRARY

liblpcisp.a can be linked to another application to add LPC programming ability.  Link liblpcisp.a
and include the headers lpcisp.h and asyncserial.h.

set where to send the report data (p=output stream)
void LPCISP_ReportStream(FILE *p);

set the report level (0=errors only, 1=minimum, 2=progress (default), 3=process debug, 4=full debug)
void LPCISP_SetReportLevel(int level);

read hex file into buffer, return pointer to buffer or NULL if error.  caller must free returned buffer.
  fileName: name of hex file to load
  start: pointer to start address (to be filled in)
  length: pointer to length of image (to be filled in)
unsigned char *LPCISP_ReadHexFile(const char *fileName,int *start,int *length);

fix up vector table.  LPC devices require vector 7 to be the 2's comp of the arithmetic sum of vectors 0-6.
  data: buffer of file image (e.g., returned by LPCISP_ReadHexFile)
void LPCISP_FixVectorTable(unsigned char *data);

open the device that we are going to communicate through, set it to 115200 baud, raw mode, return
descriptor number on success, -1 on error
int LPCISP_SERIAL_OpenDevice(const char *name)

close the device that was opened by LPCISP_SERIAL_OpenDevice
void LPCISP_SERIAL_CloseDevice(int fd);

Attempt to read the given number of bytes before timeout (uS) occurs during any read attempt.
Return the number of bytes received or -1 on error.  if maxBytes is zero will simply return zero.
 fd -- file descriptor of serial device
 buf -- pointer to buffer to be filled
 maxBytes -- max number of bytes to read
 timeout -- time to wait on any read, in microseconds
int LPCISP_SERIAL_ReadBytes(int fd,unsigned char *buf,unsigned int maxBytes,unsigned int timeOut);

Write bytes to device.  return number of bytes written (should be same as numBytes unless error).
 fd -- file descriptor
 buf -- pointer to buffer to be sent
 numBytes -- number of bytes to send
unsigned int LPCISP_SERIAL_WriteBytes(int fd,const unsigned char *buf,unsigned int numBytes);

Set the state of the DTR handshake line
void LPCISP_SERIAL_SetDTR(int fd, int state);

Set the state of the RTS handshake line
void LPCISP_SERIAL_SetRTS(int fd, int state);

flush the serial device.
void LPCISP_SERIAL_FlushDevice(int fd);

set baud rate for communication with target device
  fd - file descriptor
  baud - baud rate
  stop - number of stop bits (1 or 2)
int LPCISP_SetBaudRate(int fd, int baud, int stop);

test if range of sectors are blank.
  fd - file descriptor
  startSector - first sector in range to be checked
  endSector - last sector in range to be checked
int LPCISP_BlankCheck(int fd, int startSector, int endSector);


ChangeLog
	lpcisp -- NXP LPCxxxx ISP

	change log
	  0.0.1 -- 16 February, 2015
	    first release
	  0.0.2 -- 16 February, 2015
	    corrected clock ISP command call (divides clock by 1000)
	  0.0.3 -- 17 February, 2015
	    added LPC8xx part definitions
	  0.0.4 -- 20 March, 2015
	    added support to tbaud argument (set baud rate for terminal)
	  0.0.5 -- 6 May, 2015
	    don't enter terminal mode if an error occurred in a prior step
	  0.0.6 -- 21 May, 2015
	    added support for devices which don't uuencode their data when reading/writing RAM
	  0.0.7 -- 23 May, 2015
	    added -parts option to dump list of supported parts
		if -term is used without other operations go straight to term, don't go through Sync() first
	  0.0.8 -- 17 June, 2015
	    added support for LPC13xx parts
	    fixed handling of line termination
		added -echo command to allow enabling echo
	  0.0.9 -- 2 July, 2015
	    fixed ReadString() so it just ignores CR (works with CRLF or LF)
	  0.0.10 -- 3 July, 2015
	    set ReadString() to stop on either CR or LF -- !!! this causes problems for binary transfers !!!
	  0.0.11 -- 6 July, 2015
	  	changed ReadString() again to try to manage CRLF correctly.
	  	improved max verbose reporting
 	  0.0.12 -- 7 August, 2015
 	    don't report error if terminal is only requested operation (doesn't read device ID, was complaining it didn't know what kind of device it was)
 		added error message if no serial device was specified
	  0.0.13 -- 31 August, 2015
	    added support for LPC17xx (need non-uniform sector support for full function)
	    changed OpenDevice() to configure only necessary flags (seems to communicate more reliably)
	  0.0.14 -- 1 September, 2015
	    exit on various failures (stop processing, don't drop to terminal mode)
	    corrected flags for LPC17xx to include uuencode
	    added support for non-uniform sector sizes
	  0.0.15 -- 10 December, 2015
	    added LPC23xx to the set of supported parts
	    fixed line termination handling so all parts work correctly (uuencoded and binary transfers both work now)
	    if reset is not mapped to a pin issue a GO command to address zero instead when exiting ISP mode
	    skip over vectors (64 bytes) when verifying parts which re-map the vector table in ISP mode
	  0.0.16 -- 14 June, 2016
	    cleaned up usage of reset, added 100 ms delay after negating reset on entry to ISP mode before attempting to communicate
	  0.0.17 -- 22 June, 2016
	    added support for non-standard baud rates in OSX
	  0.0.18 -- 10 July 2016
	    added flag for HAS_UID, skip trying to read or report UID on parts which don't support it (need to fully update partDef[])
	  0.0.19 -- 29 July, 2016
	    added support for LPC18xx parts.  these parts may have multiple flash banks and have two word IDs so the part description
	    table was extended to add parameters to support the differences.  the low level functions now have the ability to write to 
	    arbitrary banks but for now only bank 0 (A) is supported.
	  0.0.20 -- 29 July, 2016
	    improved hex file handling to work properly with files which don't start at address zero
	    expanded sector map in part description to include bank and base address in addition to size (banks still not fully supported)
	  0.0.21 -- 30 July, 2016
	    added -vector argument to allow forcing of vector table patching when image does not start at zero
	  0.0.22 -- 6 December, 2017
		removed debugging code
		corrected usage instructions for -isp
		added support for LPC154x, LPC151x, LPC175x, and LPC176x (via Todd Squires)
		added flag for skipping first 256 bytes when verifying parts which re-map the bootloader into vector space in ISP mode (previously only support 64 bytes)
	  0.0.23 -- 20 December 2017
	    added support for LPC21xx and LPC22xx parts
	  0.0.24 -- 26 December 2017
	    added support for LPC822/LPC824 parts
	  0.0.25 -- 7 February 2018
	    added support for LPC546xx parts
	    corrected display of dumps which don't end on 16-byte boundary
	    added SKIP_0 flag for parts (e.g., LPC546xx) which can't read word 0 from flash
	    ensured hex file start address and number of bytes to read from flash are on word boundaries (multiple of 4 bytes)
	    added support for extended address segment records in intel hex files
	  0.0.26 -- 15 June, 2018
	    added -start command line option to allow loading file at different address than passed in hex file or 0 for binary files
	  0.0.27 -- 13 February, 2019
	    added support for Windows
	  0.0.28 -- 30 March, 2019
	    added FlushDevice() command, flush serial port after asserting reset and before communicating with target.
	  0.0.29 -- 12 September, 2019
	    added support for LPC1124 and LPC1125
	  0.0.30 -- 4 November 2019
	    change "erase" and "eraseall" to "erasemin" and "erase", respectively.  "erasemin" erases only the sectors which are affected by the
	      incoming hex file, "erase" bulk erases the entire device.
	  0.0.31 -- 20 January, 2020
	    altered code in isp.c to suppress spurious warnings from gcc 8.3.0
	    added bank support for flash erase and write
	    added "-device" option to allow operations on a specific device rather than whatever is detected (default)
	    corrected name of LPC1114[FHN33/301|FHN33/302|FHI33/302|FBD48/302|FBD100/302] in part list
	  0.0.32 -- 6 March 2020
	    corrected bug which caused verify to operate on word instead of byte boundaries
	  0.0.33 -- 27 June 2020
	    swapped major and minor boot version (were reversed)
	  0.1.0 -- 3 September 2020
	    split core functionality into a library so it can be more easily included in other applications (Unix only)
	  0.1.1 -- 26 May 2021
	    corrected naming issue which caused Windows version to fail to build
	    disabled XON/XOFF flow control on Windows driver (was causing 0x11 and 0x13 to be dropped when transfering data via binary)
	  0.1.2 -- 17 June 2021
	    corrected top-level error reporting (was incorrectly indexing into errString[])
	  0.1.3 -- 6 December 2021

main.c
/*

	lpcisp -- NXP LPCxxxx ISP

	see ChangeLog for version information.
	
	to do:
		fix ReadString() so it can't overrun the buffer (need to pass buffer length)
		properly read back first 64 or 256 bytes of LPC17xx parts

*/

#include "includes.h"

static const char
	*version="0.2.5";

static int
	fd,
	baud,
	tbaud,
	clock,
	retries,
	start,
	length,
	echo,
	hold,
	resetPin,
	ispPin,
	vector,
	burn,
	erase,
	eraseall,
	verify,
	term;

static unsigned char
	*data;

static partinfo_t
	partInfo;

static char
	*fileName,
	*deviceName;

static long
	dumpAddr,
	dumpLength;
static int
	dumpSize;

static void ReportFileInfo(int level)
{
	ReportString(level,"file information:\n");
	ReportString(level,"   file name:     %s\n",fileName);
	ReportString(level,"   start address: 0x%08x\n",start);
	ReportString(level,"   length:        0x%08x\n",length);
}

//========== option handlers ==========

static int VerboseOpt(int *argc, char ***argv)
{
	if(*argc)
	{
		LPCISP_SetReportLevel(strtol(**argv,NULL,0));
		*argc=(*argc)-1;;
		*argv=(*argv)+1;
		return(0);
	}
	return(-1);
}

static int PartsOpt(int *argc, char ***argv)
{
	DumpPartList(stdout);
	return(0);
}

static int WriteOpt(int *argc, char ***argv)
{
	burn=1;
	return(0);
}

static int EraseOpt(int *argc, char ***argv)
{
	erase=1;
	return(0);
}

static int EraseAllOpt(int *argc, char ***argv)
{
	erase=1;
	eraseall=1;
	return(0);
}

static int EchoOpt(int *argc, char ***argv)
{
	echo=1;
	return(0);
}

static int VectorOpt(int *argc, char ***argv)
{
	vector=1;
	return(0);
}

typedef struct
{
	const char
		*str;
	int
		code;
}pin_t;

static int DeviceOpt(int *argc, char ***argv)
// using the next argument, set the target device
{
	if(*argc)
	{
		deviceName=**argv;
		*argv=(*argv)+1;
		*argc=(*argc)-1;
		return(0);
	}
	return(-1);
	
}

static int GetCtrlArg(const char *arg)
{
	int
		i;
	const pin_t
		pinMap[]=
		{
			{	"dtr",	PIN_DTR		},
			{	"ndtr",	PIN_nDTR	},
			{	"rts",	PIN_RTS		},
			{	"nrts",	PIN_nRTS	},
		};
	for(i=0;i<ARRAY_SIZE(pinMap);i++)
	{
		if(strcmp(arg,pinMap[i].str)==0)
		{
			return(pinMap[i].code);
		}
	}
	return(-1);
}

static int ResetOpt(int *argc, char ***argv)
// using the next argument, map the reset control to specified serial device pin.
{
	int
		pin;

	if(*argc)
	{
		if((pin=GetCtrlArg(**argv))>=0)
		{
			resetPin=pin;
			*argv=(*argv)+1;
			*argc=(*argc)-1;
			return(0);
		}
	}
	return(-1);
}

static int ISPOpt(int *argc, char ***argv)
// using the next argument, map the ISP control to specified serial device pin.
{
	int
		pin;

	if(*argc)
	{
		if((pin=GetCtrlArg(**argv))>=0)
		{
			ispPin=pin;
			*argv=(*argv)+1;
			*argc=(*argc)-1;
			return(0);
		}
	}
	return(-1);
}

static int HoldOpt(int *argc, char ***argv)
{
	hold=1;
	return(0);
}

static int BaudOpt(int *argc, char ***argv)
// using the next argument get the baud rate
{
	if(*argc)
	{
		if((baud=strtol(**argv,NULL,0)))
		{
			*argv=(*argv)+1;
			*argc=(*argc)-1;
			return(0);
		}
	}
	return(-1);
}

static int TBaudOpt(int *argc, char ***argv)
// using the next argument get the terminal baud rate
{
	if(*argc)
	{
		if((tbaud=strtol(**argv,NULL,0)))
		{
			*argv=(*argv)+1;
			*argc=(*argc)-1;
			return(0);
		}
	}
	return(-1);
}

static int ClockOpt(int *argc, char ***argv)
// using the next argument get the clock rate
{
	if(*argc)
	{
		if((clock=strtol(**argv,NULL,0)))
		{
			*argv=(*argv)+1;
			*argc=(*argc)-1;
			return(0);
		}
	}
	return(-1);
}


static int RetryOpt(int *argc, char ***argv)
// using the next argument get the number of sync retries
{
	if(*argc)
	{
		if((retries=strtol(**argv,NULL,0)))
		{
			*argv=(*argv)+1;
			*argc=(*argc)-1;
			return(0);
		}
	}
	return(-1);
}

static int VerifyOpt(int *argc, char ***argv)
{
	verify=1;
	return(0);
}

static int TermOpt(int *argc, char ***argv)
{
	term=1;
	return(0);
}

static int DumpOpt(int *argc, char ***argv)
{
	if(*argc>=2)
	{
		dumpAddr=strtol(**argv,NULL,0);
		*argv=(*argv)+1;
		dumpLength=strtol(**argv,NULL,0);
		*argv=(*argv)+1;
		*argc=(*argc)-2;
		dumpSize=1;
		if((*argc>1)&&(**argv[0]!='-'))
		{
			dumpSize=strtol(**argv,NULL,0);
			*argv=(*argv)+1;
			*argc=(*argc)-1;
		}
		if((dumpSize!=1)&&(dumpSize!=2)&&(dumpSize!=4))
		{
			dumpSize=1;
		}
		return(0);
	}
	return(-1);
}


static int HexFileOpt(int *argc, char ***argv)
{
	if(*argc)
	{
		if(data)
		{
			// throw out a previously-loaded buffer if one exists
			free(data);
		}

		// read in a hex file
		fileName=**argv;
		data=LPCISP_ReadHexFile(**argv,&start,&length);
		if(data)
		{
			ReportFileInfo(REPORT_INFO);
			*argc=(*argc)-1;;
			*argv=(*argv)+1;
			return(0);
		}
	}
	return(-1);
}

static int StartOpt(int *argc, char ***argv)
{
	if(*argc)
	{
		start=strtol(**argv,NULL,0);
		*argc=(*argc)-1;
		*argv=(*argv)+1;
		return(0);
	}
	return(-1);
}

static int BinFileOpt(int *argc, char ***argv)
{
	FILE
		*fp;
	int
		retval;

	retval=-1;
	if(*argc)
	{
		// save file name
		fileName=**argv;
		*argc=(*argc)-1;;
		*argv=(*argv)+1;
		// throw out a previously-loaded buffer if one exists
		if(data)
		{
			free(data);
		}

		// get size (and existence) of file, allocate a buffer, read it in
		if((fp=fopen(fileName,"rb")))
		{
			// get file size
			fseek(fp,0L,SEEK_END);
			if((length=ftell(fp)))
			{
				rewind(fp);
				start=0;
				data=(unsigned char *)malloc(length);
				if(data)
				{
					if((fread(data,1,length,fp)==length))
					{
						ReportFileInfo(REPORT_INFO);
						retval=0;
					}
					else
					{
						ReportString(REPORT_ERROR,"failed to read %d bytes from \"%s\": %s\n",length,fileName,strerror(errno));
					}
				}
				else
				{
					ReportString(REPORT_ERROR,"failed to allocate memory: %s\n",strerror(errno));
				}
			}
			else
			{
				ReportString(REPORT_ERROR,"zero-length file: %s\n",fileName);
			}
			fclose(fp);
		}
		else
		{
			ReportString(REPORT_ERROR,"failed to open \"%s\" for reading: %s\n",fileName,strerror(errno));
		}
	}
	return(retval);
}

//========== token handling ==========

typedef int (funct_t)(int *argc, char ***argv);

typedef struct
{
	const char
		*token;
	funct_t
		*funct;
	const char
		*helpStr;
}token_t;

static token_t
	tokenList[]=
	{
			// token name	function		help string
		{	"-device",		DeviceOpt,		" devicename           specify an LPC device.  default is to attempt to program whatever is attached."						},
		{	"-reset",		ResetOpt,		" ctrl                  map reset to a serial control signal.  ctrl may be one of dtr, ndtr, rts, or nrts (n=inverted)"		},
		{	"-isp",			ISPOpt,			" ctrl                    map ISP to a serial control signal.  ctrl may be one of dtr, ndtr, rts, or nrts (n=inverted)"		},
		{	"-hold",		HoldOpt,		"                        assert ISP through programming sequence (default is to negate at initialization)"					},
		{	"-baud",		BaudOpt,		" rate                   set the baud rate for ISP mode (default=115200)"													},
		{	"-tbaud",		TBaudOpt,		" rate                  set the baud rate for the terminal (default=115200)"												},
		{	"-clock",		ClockOpt,		" rate                  set the clock rate in kHz (default=12000)"															},
		{	"-retry",		RetryOpt,		" num                   set number of retries when synchronizing (default=25)"												},
		{	"-bin",			BinFileOpt,		" filename                specify name of binary file to load"																},
		{	"-hex",			HexFileOpt,		" filename                specify name of hex file to load"																	},
		{	"-start",		StartOpt,		" address               override start address (specify AFTER hex or binary file)"											},
		{	"-erasemin",	EraseOpt,		"                    erase only required sectors on target device (may fail if code is read-protected)"						},
		{	"-erase",		EraseAllOpt,	"                       erase all sectors on target device"																	},
		{	"-echo",		EchoOpt,		"                        enable echo from target (default is no echo)"														},
		{	"-vector",		VectorOpt,		"                      patch vector 7 to 2's complement of the sum of vectors 0 through 6 regardless of address of image"	},
		{	"-write",		WriteOpt,		"                       write image to target device"																		},
		{	"-verify",		VerifyOpt,		"                      verify image"																						},
		{	"-dump",		DumpOpt,		" address length [size]  read from length bytes from memory starting at address, dump to console"							},
		{	"-term",		TermOpt,		"                        enter terminal mode after other operations"														},
		{	"-verbose",		VerboseOpt,		" level               set reporting level (0=errors only, 1=minimum, 2=progress (default), 3=process debug, 4=full debug)"	},
		{	"-parts",		PartsOpt,		"                       list supported devices"																				},
	};

static void Usage(const char *progName)
{
	int
		i;

	ReportString(REPORT_ERROR,"%s: ver. %s (c) 2015-2021 Cosmodog, Ltd.\n",progName,version);
	ReportString(REPORT_ERROR,"usage:\n");
	ReportString(REPORT_ERROR,"  %s [options] device\n",progName);
	ReportString(REPORT_ERROR,"where\n");
	ReportString(REPORT_ERROR,"  device is the serial device (/dev/ttyS0, etc.)\n");
	ReportString(REPORT_ERROR,"options:\n");
	for(i=0;i<ARRAY_SIZE(tokenList);i++)
	{
		ReportString(REPORT_ERROR,"  %s%s\n",tokenList[i].token,tokenList[i].helpStr);
	}
}

static int MatchDeviceName(char *targetName, const char *detectedName)
// compare requested device name to detected device, return TRUE if targetName is undefined or names match, FALSE if not.
// note: name must match the name from the partDef array in partdesc.c.  at some point it might be nice to make the
// match more flexible, so, for instance, "LPC1111FHN33/101" should match "LPC1111FHN33/[101|102]".
{
	int
		i;

	if(targetName)
	{
		for(i=0;i<strlen(targetName);i++)
		{
			// convert to upper case to make comparison case-insensitive
			targetName[i]=toupper(targetName[i]);
		}
	}
	return(!targetName||(strcmp(targetName,detectedName)==0));
}

//========== main ==========
int main(int argc,char *argv[])
{
	int
		i,
		j,
		lineLen,
		fail;
	char
		*progName;
	const char
		*errString[]=
		{
			"Success",
			"Failed to set baud rate",
			"Failed to sync with target",
			"Communication error",
			"Failed to get device info from target",
			"Unknown LPC device",
		};
	lpcispcfg_t
		cfg;

	fail=0;
	LPCISP_ReportStream(stdout);
#if defined __UNIX__
	progName=basename(*argv);
#else
	progName=*argv;
#endif
	argc--;
	argv++;
	if(argc)
	{
		// set the defaults.  may be changed by command line arguments
		fd=-1;
		baud=115200;
		tbaud=115200;
		clock=12000;
		retries=25;
		resetPin=PIN_NONE;
		ispPin=PIN_NONE;
		hold=0;
		LPCISP_SetReportLevel(REPORT_INFO);
		erase=0;
		eraseall=0;
		echo=0;
		vector=0;
		burn=0;
		verify=0;
		term=0;
		start=0;
		length=0;
		dumpAddr=-1;
		dumpLength=-1;
		deviceName=(char *)NULL;
		data=(unsigned char *)NULL;

		// parse command line arguments
		while(!fail&&argc&&(*argv[0]=='-'))
		{
			// option, process it
			for(i=0;i<ARRAY_SIZE(tokenList);i++)
			{
				if(strcmp(tokenList[i].token,*argv)==0)
				{
					// skip over the token and call the function which may then consume more arguments.
					argc--;
					argv++;
					fail=(tokenList[i].funct(&argc,&argv)<0);
					if(fail)
					{
						ReportString(REPORT_ERROR,"invalid arguments\n");
					}
					break;
				}
			}
			if(!fail)
			{
				fail=(i==ARRAY_SIZE(tokenList));
				if(fail)
				{
					ReportString(REPORT_ERROR,"unknown option \"%s\"\n",*argv);
				}
			}
			if(!fail&&(start&0x03))
			{
				ReportString(REPORT_ERROR,"start address (0x%08x) must be on a word boundary (multiple of four)\n",start);
				fail=1;
			}
		}

		if(!fail&&argc)
		{
			// expect a device name (e.g., /dev/ttyS0) as the last argument.  open it and try to communicate with
			// a target device and retrieve device information.
			fd=LPCISP_SERIAL_OpenDevice(*argv);
			fail=(fd<0);
			if(!fail)
			{
				// if term was the only action requested skip the synchronization step and drop directly to terminal
				if(erase||data||(dumpLength>0)||!term)
				{
					fail=LPCISP_Sync(fd,&cfg,baud,clock,retries,echo,hold,resetPin,ispPin,&partInfo);
					if(!fail)
					{
						// NULL deviceName matches anything
						fail=!MatchDeviceName(deviceName,partInfo.name);
						if(!fail)
						{
							ReportPartInfo(REPORT_INFO,&partInfo);
						}
						else
						{
							ReportString(REPORT_ERROR,"part mismatch, wanted \"%s\", detected \"%s\"\n",deviceName,partInfo.name);
						}
					}
					else
					{
						ReportString(REPORT_ERROR,"failed: %s\n",errString[-fail]);
						if(fail==-ISP_ERR_UNKNOWN_DEV)
						{
							ReportString(REPORT_ERROR,"  id: %08x\n",partInfo.id[0]);
						}
					}
				}
			}
			else
			{
				ReportString(REPORT_ERROR,"failed to open \"%s\"\n",*argv);
			}
		}

		// if a device was opened try to do operations on it
		if(!fail&&(fd>=0))
		{
			if(erase||data||(dumpLength>0))
			{
				// perform operations only if part was in our database
				if(partInfo.numSectors)
				{
					if(erase)
					{
						int
							startSector,
							endSector;

						if(eraseall||(length==0))
						{
							// asked to erase all or no file provided, erase all sectors
							startSector=0;
							endSector=partInfo.numSectors-1;
						}
						else
						{
							// find sectors which contain start address and end address
							startSector=LPCISP_GetSectorAddr(start,&partInfo);
							endSector=LPCISP_GetSectorAddr(start+length,&partInfo);
						}
						fail=(LPCISP_Erase(fd,&cfg,startSector,endSector,0,&partInfo)<0);	// @@@ only support bank zero for now
					}

					// if data image was loaded we can write it to flash and/or verify it
					if(!fail&&data)
					{
						// if image starts at 0x0000 or forced fix up the vector table.
						if(vector||(start==0))
						{
							LPCISP_FixVectorTable(data);
						}

						// write image to flash
						if(burn)
						{
							fail=(LPCISP_WriteToFlash(fd,&cfg,data,start,length,&partInfo)<0);
						}

						// compare image against flash
						if(!fail&&verify)
						{
							fail=(LPCISP_VerifyFlash(fd,&cfg,data,start,length,&partInfo)<0);
						}
						free(data);
					}

					if(dumpLength>0)
					{
						i=0;
						data=(unsigned char *)malloc(dumpLength);
						if(data)
						{
							// dumpAddr and dumpLength must be word-aligned
							dumpAddr&=~3;
							dumpLength=(dumpLength+3)&~3;
							fail=(LPCISP_ReadFromTarget(fd,&cfg,data,dumpAddr,dumpLength,&partInfo)<0);
							if(!fail)
							{
								while(i<dumpLength)
								{
									lineLen=16-(dumpAddr&0x0f);
									ReportString(REPORT_MINIMUM,"%08x: ",dumpAddr&~0x0f);
									for(j=0;j<16-lineLen;j++)
									{
										ReportString(REPORT_MINIMUM,"   ");
									}
									for(j=0;j<lineLen;j+=dumpSize)
									{
										if((i+j)<dumpLength)
										{
											switch(dumpSize)
											{
												case 4:
													ReportString(REPORT_MINIMUM,"%02x%02x%02x%02x ",(unsigned char)data[i+j+3],(unsigned char)data[i+j+2],(unsigned char)data[i+j+1],(unsigned char)data[i+j]);
													break;
												case 2:
													ReportString(REPORT_MINIMUM,"%02x%02x ",(unsigned char)data[i+j+1],(unsigned char)data[i+j]);
													break;
												default:
													ReportString(REPORT_MINIMUM,"%02x ",(unsigned char)data[i+j]);
													break;
											}
										}
										else
										{
											// end of data on last line, pad so ASCII lines up
											ReportString(REPORT_MINIMUM,dumpSize==1?"   ":(dumpSize==2?"     ":"         "));
										}
									}
									ReportString(REPORT_MINIMUM," - ");
									for(j=0;j<16-lineLen;j++)
									{
										ReportString(REPORT_MINIMUM," ");
									}
									for(j=0;(j<lineLen)&&((i+j)<dumpLength);j++)
									{
										ReportString(REPORT_MINIMUM,"%c",(data[i+j]>=' '&&data[i+j]<='~')?data[i+j]:'.');
									}
									i+=lineLen;
									dumpAddr+=lineLen;
									ReportString(REPORT_MINIMUM,"\n");
								}
							}
							free(data);
						}
					}
				}
				else
				{
					ReportString(REPORT_ERROR,"ERROR: unknown part, cannot perform operations on it\n");
				}
			}
			LPCISP_ExitISPMode(fd,&cfg,&partInfo);

			if(!fail&&term)
			{
				// go into terminal mode with current line settings
				if(tbaud!=baud)
				{
					fail=(LPCISP_SERIAL_ChangeBaudRate(fd,tbaud)<0);
				}
				if(!fail)
				{
					fail=!Terminal(fd,&cfg);
				}
			}

			LPCISP_SERIAL_CloseDevice(fd);
		}
		if(fd<0)
		{
			ReportString(REPORT_ERROR,"ERROR: no serial device specified\n");
			fail=1;
		}
	}
	else
	{
		Usage(progName);
	}
	return(fail?-1:0);
}
